Revised Docker flow run tutorial (#6612)
PrefectHQ/prefect: docs/tutorials/docker-flow-runner.md

---
description: Learn how to build Prefect deployments that create flow runs in Docker containers.
tags:
    - Docker
    - containers
    - orchestration
    - infrastructure
    - deployments
    - tutorial
---

# Running flows with Docker

> In the [Deployments](/tutorials/deployments/) and [Storage and Infrastructure](/tutorials/storage/) tutorials, we looked at creating configuration that enables creating flow runs via the API and with code that was uploaded to a remotely accessible location.  

Reviewer: "/tutorials/storage/" 404ed for me

> In this tutorial, we'll further configure the deployment so flow runs are executed not locally, but in a the "remote" execution environment of a Docker container. 

Reviewer: ```suggestion
In this tutorial, we'll further configure the deployment so that flow runs are executed in the execution environment of a Docker container. We'll run our Docker instance locally, but you can extend this tutorial to run it on remote machines.
```

Reviewer: My wording might not be perfect - but in this example we are running Docker locally. I don't want folks to get confused thinking that it is remote on AWS or something. We do have examples of that in Discourse and might be worth putting in the Docs at a later date.

In this tutorial we'll: 

- Configure a Docker Container infrastructure block that enables creating flow runs in a container.
- Build and apply a new `log_flow.py` deployment that uses the new infrastructure block.
- Create a flow run from this deployment that spins up a Docker container and executes, logging a message.

## Prerequisites

To run a deployed flow in a Docker container, you'll need the following:

- We'll use the flow script and deployment from the [Deployments](/tutorials/deployments/) tutorial. 
- We'll also use the remote storage block created in the [Storage and Infrastructure](/tutorials/storage/) tutorial.

> - You must run a standalone Prefect Orion API server (`prefect orion start`) or Prefect Cloud.

Reviewer: ```suggestion
- You must run a standalone Prefect Orion API server (`prefect orion start`) or use Prefect Cloud.
```

[Docker Desktop](https://www.docker.com/products/docker-desktop) works fine for local testing if you don't already have Docker Engine configured in your environment.

!!! note "Run Prefect Orion server"
    This tutorial assumes you're already running a Prefect Orion server with `prefect orion start`, as described in the [Deployments](/tutorials/deployments/#run-a-prefect-orion-server) tutorial. 
    
    If you shut down the server from a previous tutorial, you can start it again by opening another terminal session and starting the Prefect Orion server with the `prefect orion start` CLI command.

## Create an infrastructure block

Most users will find it easiest to configure new infrastructure blocks through the Prefect Orion or Prefect Cloud UI. 

You can see any previously configured storage blocks by opening the Prefect UI and navigating to the **Blocks** page. To create a new infrastructure block, select the **+** button on this page. Prefect displays a page of available block types. Select **run-infrastructure** from the **Capability** list to filter just the infrastructure blocks.

> ![Viewing a list of infrastructure block types in the Prefect UI](/img/tutorials/infrastructure-blocks.png)

Reviewer: This 404s for me.

Author: These images are from the storage tutorial PR. They will work when that is merged.

Use these base blocks to create your own infrastructure blocks containing the settings needed to run flows in your environment.

For this tutorial, find the **Docker Container** block, then select **Add +** to see the options for a Docker infrastructure block.

> ![Viewing a list of infrastructure block types in the Prefect UI](/img/tutorials/docker-infrastructure.png)

Reviewer: also 404s for me

## Edit the deployment in the UI

`log_flow` exepcts a runtime parameter for its greeting, and we didn't provide one as part of this deployment yet. We could edit `log-flow-docker-deployment.yaml` to add a parameter and apply the edited YAML to update the deployment on the API.

Instead, let's edit the deployment through the Prefect UI. Select **log-flow/log-flow-docker** to see the deployment's details.

![Viewing the Docker deployment details in the Prefect UI](/img/tutorials/docker-deployment-details.png)

Select the menu next to **Run**, then select **Edit** to edit the deployment.

Scroll down to the **Parameters** section and provide a value for the `name` parameter. We used "Ford Prefect" here. 

![Editing the Docker deployment details in the Prefect UI](/img/tutorials/edit-docker-deployment.png)

Select **Save** to save these changes to the deployment.

## Create a flow run in Docker

When you create flow runs from this deployment, the agent pulls the default Prefect Docker container, `pip installs` the prerequisites we specified, retrieves the flow script from remote storage, and starts the Prefect engine to execute the flow run.

> Let's create a flow run for this deployment. The flow run will execute in a Docker container rather than in your local development environment.

Reviewer: ```suggestion
Let's create a flow run for this deployment. The flow run will execute in a Docker container on your local machine.
```

!!! note "Run a Prefect agent"

>     This tutorial assumes you're already running a Prefect agent with `prefect agent start`, as described in the [Deployments](/tutorials/deployments/#agents-and-work-queues) tutorial. 

Reviewer: I think the intention here is to run the agent from within the Docker container, no? If so, I would tell folks to shut down their other agent so that things aren't confusing. Then start an agent in the Docker container.

Author: No in this simple tutorial, the local agent is spinning up the docker container, executes the flow run in that container, then shuts down the container. Getting an agent running in a docker container isn't really related to infrastructure, it's a different problem to solve and we'll get to that next.
